packages/nextcloud/lib/src/webdav/path_uri.dart

```diff
@@ -0,0 +1,76 @@
+import 'package:collection/collection.dart';
+import 'package:meta/meta.dart';
+
+/// A [Uri] that only exposes the path handling.
+@immutable
+class PathUri {
+  /// Creates a new path URI.
+  ///
+  /// Only one of [path] and [pathSegments] must be given.
+  PathUri({final String? path, final List<String>? pathSegments})
+      : assert(
+          (path == null) != (pathSegments == null) || (path == null && pathSegments == null),
+          'Only one of path and pathSegments must be given',
+        ),
+        _uri = Uri(
+          path: path,
+          pathSegments: pathSegments,
+        ).normalizePath();
+
+  final Uri _uri;
```

Comment:
I don't like having an actual `_Uri` backing ours. Both default implementations of the `Uri` interface `_Uri` and `_SimpleUri` only store the path in a String and compute the segments on the fly.

As we most often use the segments we should use a List (or BuiltList) to back our implementation. This should improve the performance.